semigroups/semiring.py

```diff
@@ -6,76 +6,323 @@
 # FIXME this is only partially complete
 
 class Semiring:
+    r"""
+    A semiring is a set :math:`R`, together with two binary operations,
+    :math:`+, \times`, such that :math:`(R, +)` is a commutative monoid,
+    with identity called 0, :math:`(R\backslash\{0\},\times)` is a 
```

Comment:
Put the definition in words, then define distributive and annihilate after. Also use words not symbols for "for all" etc.